src/main/scala/io/ticofab/AwsSigner.scala

```diff
@@ -167,6 +168,21 @@ class AwsSigner(credentialsProvider: AWSCredentialsProvider,
       result += (X_AMZ_DATE -> now.format(DATE_FORMATTER))
     }
 
+    val payloadSignature = payload.map(p => toBase16(hash(p)))
+
+    val contentSignature = result.get(CONTENT_SHA256_HEADER) match {
+      case Some(signature) =>
+        if (payloadSignature.exists(_ != signature))
+          throw new IllegalArgumentException(s"Request already contain $CONTENT_SHA256_HEADER with hash " +
+            s"different than hash of provided payload.")
```

Comment:
I am in favour of handling errors in a functional way with constructs like Try or Either. For instance, this match case could return either a Success or a Failure, or some default value. We can do better than throwing exceptions, see for instance [here](http://danielwestheide.com/blog/2012/12/26/the-neophytes-guide-to-scala-part-6-error-handling-with-try.html).